SyncTests/TestBookmarkTreeMerging.swift

```diff
@@ -290,6 +290,114 @@ class TestBookmarkTreeMerging: FailFastTestCase {
         XCTAssertEqual(Set(BookmarkRoots.Real), result.overrideCompletion.processedLocalChanges)
     }
 
+    func testMergingInconsistentBufferWithMissingValues() {
+        guard let bookmarks = self.getSyncableBookmarks("H") else {
+            XCTFail("Couldn't get bookmarks.")
+            return
+        }
+
+        // Setup a buffer with a folder that points to a bookmark but doesn't actually contain said bookmark.
+        let bufferDate = NSDate.now()
+        let changedBufferRecords = [
+            BookmarkMirrorItem.folder("folderA", modified: bufferDate, hasDupe: false, parentID: BookmarkRoots.RootGUID, parentName: "", title: "folderA", description: nil, children: ["bookmarkA"])
+        ]
+
+        bookmarks.applyRecords(changedBufferRecords).succeeded()
+
+        let uploader = MockUploader()
+        let storer = uploader.getStorer()
+        let applier = MergeApplier(buffer: bookmarks, storage: bookmarks, client: storer, greenLight: { true })
+        let result = applier.go().value
+        XCTAssertFalse(result.isSuccess)
```

Comment:
We have a test helper:

```
extension Deferred where T: Succeedable {
    func succeeded() {
        XCTAssertTrue(self.value.isSuccess)
    }
}
```

Extend that to have `.failed`, too, and use it here and throughout.